src/app/clusters/software-diagnostics-server/software-diagnostics-logic.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/clusters/software-diagnostics-server/software-diagnostics-logic.h>
+
+#include <app/server-cluster/DefaultServerCluster.h>
+#include <lib/support/CodeUtils.h>
+
+#include <clusters/SoftwareDiagnostics/Commands.h>
+#include <clusters/SoftwareDiagnostics/Ids.h>
+#include <clusters/SoftwareDiagnostics/Metadata.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace {
+
+/// Wrapper around `DeviceLayer::GetDiagnosticDataProvider::GetThreadMetrics` that ensures
+/// that `ReleaseThreadMetrics` is always called on the underlying value.
+class AutoFreeThreadMetrics
+{
+public:
+    AutoFreeThreadMetrics(DeviceLayer::DiagnosticDataProvider & provider) : mProvider(provider) {}
+    ~AutoFreeThreadMetrics()
+    {
+        if (mMetrics != nullptr)
+        {
+            mProvider.ReleaseThreadMetrics(mMetrics);
+        }
+    }
+
+    CHIP_ERROR ReadThreadMetrics()
+    {
+        VerifyOrReturnError(mMetrics == nullptr, CHIP_ERROR_INCORRECT_STATE);
+        return mProvider.GetThreadMetrics(&mMetrics);
```

Comment:
Does the callee guarantee that the outparam is set to something sane on failure?

If it's not documented in the API there (and I bet it's not), I would not assume it happens.  Certainly the old code did not assume anything like that.